File: src/main/java/com/github/youngerdryas89/moviescraper/model/MovieFactory.java
package com.github.youngerdryas89.moviescraper.model;

import java.util.ArrayList;
import com.github.youngerdryas89.moviescraper.model.dataitem.Actor;
import com.github.youngerdryas89.moviescraper.model.dataitem.Director;
import com.github.youngerdryas89.moviescraper.model.dataitem.Genre;
import com.github.youngerdryas89.moviescraper.model.dataitem.ID;
import com.github.youngerdryas89.moviescraper.model.dataitem.MPAARating;
import com.github.youngerdryas89.moviescraper.model.dataitem.OriginalTitle;
import com.github.youngerdryas89.moviescraper.model.dataitem.Outline;
import com.github.youngerdryas89.moviescraper.model.dataitem.Plot;
import com.github.youngerdryas89.moviescraper.model.dataitem.Rating;
import com.github.youngerdryas89.moviescraper.model.dataitem.ReleaseDate;
import com.github.youngerdryas89.moviescraper.model.dataitem.Runtime;
import com.github.youngerdryas89.moviescraper.model.dataitem.Set;
import com.github.youngerdryas89.moviescraper.model.dataitem.SortTitle;
import com.github.youngerdryas89.moviescraper.model.dataitem.Studio;
import com.github.youngerdryas89.moviescraper.model.dataitem.Tag;
import com.github.youngerdryas89.moviescraper.model.dataitem.Tagline;
import com.github.youngerdryas89.moviescraper.model.dataitem.Thumb;
import com.github.youngerdryas89.moviescraper.model.dataitem.Title;
import com.github.youngerdryas89.moviescraper.model.dataitem.Top250;
import com.github.youngerdryas89.moviescraper.model.dataitem.Trailer;
import com.github.youngerdryas89.moviescraper.model.dataitem.Votes;
import com.github.youngerdryas89.moviescraper.model.dataitem.Year;

public class MovieFactory {

	/**
	 * @return A movie object with every field initialized to either blank (for things like string values) or having zero values (array/list objects)
	 */
	public static Movie createEmptyMovie() {

		ArrayList<Actor> actors = new ArrayList<>();

		ArrayList<Director> directors = new ArrayList<>();

		Thumb[] fanart = new Thumb[0];
		Thumb[] extraFanart = new Thumb[0];

		ArrayList<Genre> genres = new ArrayList<>();

		ArrayList<Tag> tags = new ArrayList<>();

		ID id = ID.BLANK_ID;
		MPAARating mpaa = MPAARating.BLANK_RATING;
		OriginalTitle originalTitle = OriginalTitle.BLANK_ORIGINALTITLE;
		Outline outline = Outline.BLANK_OUTLINE;
		Plot plot = Plot.BLANK_PLOT;
		Thumb[] posters = new Thumb[0];

		Rating rating = Rating.BLANK_RATING;
		ReleaseDate releaseDate = ReleaseDate.BLANK_RELEASEDATE;
		Runtime runtime = Runtime.BLANK_RUNTIME;
		Set set = Set.BLANK_SET;
		SortTitle sortTitle = SortTitle.BLANK_SORTTITLE;
		Studio studio = Studio.BLANK_STUDIO;
		Tagline tagline = Tagline.BLANK_TAGLINE;
		Title title = new Title("");
		Top250 top250 = Top250.BLANK_TOP250;
		Trailer trailer = Trailer.BLANK_TRAILER;
		Votes votes = Votes.BLANK_VOTES;
		Year year = Year.BLANK_YEAR;

		return new Movie(actors, directors, fanart, extraFanart, genres, tags, id, mpaa, originalTitle, outline, plot, posters, rating, releaseDate, runtime, set, sortTitle, studio, tagline, title,
		        top250, trailer, votes, year);
	}

}

File: src/main/java/com/github/youngerdryas89/moviescraper/model/MovieFactory.kt
package com.github.youngerdryas89.moviescraper.model

import com.github.youngerdryas89.moviescraper.controller.xmlserialization.KodiXmlMovieBean
import com.github.youngerdryas89.moviescraper.model.dataitem.*
import org.apache.commons.io.IOUtils
import java.io.File
import java.io.FileInputStream
import java.io.IOException

object MovieFactory {
    /**
     * Create a movie by reading in a values from a nfo file created by previously scraping the movie and then writing the metadata out to the file
     * 
     * @param nfoFile
     * @throws IOException
     */
    @JvmStatic
	@Throws(IOException::class)
    fun createMovieFromNfo(nfoFile: File): Movie? {
        var movieFromNfo: Movie? = null
        FileInputStream(nfoFile).use { fisTargetFile ->
            var targetFileStr = IOUtils.toString(fisTargetFile, "UTF-8")
            // Sometimes there's some junk before the prolog tag. Find the start of the XML content and use that.
            val xmlStartIndex = targetFileStr.indexOf("<?xml")
            if (xmlStartIndex > 0) {
                targetFileStr = targetFileStr.substring(xmlStartIndex)
            } else if (xmlStartIndex == -1) {
                // If there's no XML tag at all, we can't parse it.
                return null
            }
            val xmlMovieBean = KodiXmlMovieBean.makeFromXML(targetFileStr)
            if (xmlMovieBean != null) {
                movieFromNfo = xmlMovieBean.toMovie()
            }
            return movieFromNfo
        }
    }

    @JvmStatic
	fun createEmptyMovie(): Movie {
        val actors = ArrayList<Actor?>()
        val directors = ArrayList<Director?>()
        val genres = ArrayList<Genre?>()
        val tags = ArrayList<Tag?>()

        val fanart = arrayOfNulls<Thumb>(0)
        val extraFanart = arrayOfNulls<Thumb>(0)
        val posters = arrayOfNulls<Thumb>(0)

        val id = ID.BLANK_ID
        val mpaa = MPAARating.BLANK_RATING
        val originalTitle = OriginalTitle.BLANK_ORIGINALTITLE
        val outline = Outline.BLANK_OUTLINE
        val plot = Plot.BLANK_PLOT
        val rating = Rating.BLANK_RATING
        val releaseDate = ReleaseDate.BLANK_RELEASEDATE
        val runtime = Runtime.BLANK_RUNTIME
        val series = Series.BLANK_SERIES
        val sortTitle = SortTitle.BLANK_SORTTITLE
        val studio = Studio.BLANK_STUDIO
        val tagline = Tagline.BLANK_TAGLINE
        val title = Title.BLANK_TITLE
        val top250 = Top250.BLANK_TOP250
        val trailer = Trailer.BLANK_TRAILER
        val votes = Votes.BLANK_VOTES
        val year = Year.BLANK_YEAR

        return Movie(
            actors,
            directors,
            fanart,
            extraFanart,
            genres,
            tags,
            id,
            mpaa,
            originalTitle,
            outline,
            plot,
            posters,
            rating,
            releaseDate,
            runtime,
            series,
            sortTitle,
            studio,
            tagline,
            title,
            top250,
            trailer,
            votes,
            year
        )
    }
}
